File: RaspberryPi_JetsonNano/python/lib/waveshare_epd/epd4in2b_V2.py
# *****************************************************************************
# * | File        :	  epd4in2bc.py
# * | Author      :   Waveshare team
# * | Function    :   Electronic paper driver
# * | Info        :
# *----------------
# * | This version:   V4.1
# * | Date        :   2022-08-10
# # | Info        :   python demo
# -----------------------------------------------------------------------------
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

import logging
from . import epdconfig

# Display resolution
EPD_WIDTH       = 400
EPD_HEIGHT      = 300

logger = logging.getLogger(__name__)

class EPD:
    def __init__(self):
        self.reset_pin = epdconfig.RST_PIN
        self.dc_pin = epdconfig.DC_PIN
        self.busy_pin = epdconfig.BUSY_PIN
        self.cs_pin = epdconfig.CS_PIN
        self.width = EPD_WIDTH
        self.height = EPD_HEIGHT

    # Hardware reset
    def reset(self):
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(200) 
        epdconfig.digital_write(self.reset_pin, 0)
        epdconfig.delay_ms(5)
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(200)   

    def send_command(self, command):
        epdconfig.digital_write(self.dc_pin, 0)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([command])
        epdconfig.digital_write(self.cs_pin, 1)

    def send_data(self, data):
        epdconfig.digital_write(self.dc_pin, 1)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([data])
        epdconfig.digital_write(self.cs_pin, 1)

    # send a lot of data   
    def send_data2(self, data):
        epdconfig.digital_write(self.dc_pin, 1)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte2(data)
        epdconfig.digital_write(self.cs_pin, 1)
        
    def ReadBusy(self):
        logger.debug("e-Paper busy")
        self.send_command(0x71)
        while(epdconfig.digital_read(self.busy_pin) == 0): # 0: idle, 1: busy
            self.send_command(0x71)
            epdconfig.delay_ms(20)
        logger.debug("e-Paper busy release")
            
    def init(self):
        if (epdconfig.module_init() != 0):
            return -1
            
        self.reset()
        
        self.send_command(0x04); 
        self.ReadBusy();

        self.send_command(0x00);
        self.send_data(0x0f);
        
        return 0

    def getbuffer(self, image):
        # logger.debug("bufsiz = ",int(self.width/8) * self.height)
        buf = [0xFF] * (int(self.width/8) * self.height)
        image_monocolor = image.convert('1')
        imwidth, imheight = image_monocolor.size
        pixels = image_monocolor.load()
        # logger.debug("imwidth = %d, imheight = %d",imwidth,imheight)
        if(imwidth == self.width and imheight == self.height):
            logger.debug("Horizontal")
            for y in range(imheight):
                for x in range(imwidth):
                    # Set the bits for the column of pixels at the current position.
                    if pixels[x, y] == 0:
                        buf[int((x + y * self.width) / 8)] &= ~(0x80 >> (x % 8))
        elif(imwidth == self.height and imheight == self.width):
            logger.debug("Vertical")
            for y in range(imheight):
                for x in range(imwidth):
                    newx = y
                    newy = self.height - x - 1
                    if pixels[x, y] == 0:
                        buf[int((newx + newy*self.width) / 8)] &= ~(0x80 >> (y % 8))
        return buf

    def display(self, imageblack, imagered):
        self.send_command(0x10)
        self.send_data2(imageblack)
        
        self.send_command(0x13)
        self.send_data2(imagered)
        
        self.send_command(0x12) 
        epdconfig.delay_ms(20)
        self.ReadBusy()
        
    def Clear(self):
        if self.width%8 == 0:
            linewidth = int(self.width/8)
        else:
            linewidth = int(self.width/8) + 1

        self.send_command(0x10)
        self.send_data2([0xff] * int(self.height * linewidth))
            
        self.send_command(0x13)
        self.send_data2([0xff] * int(self.height * linewidth))
        
        self.send_command(0x12) 
        epdconfig.delay_ms(20)
        self.ReadBusy()

    def sleep(self):
        self.send_command(0X50)
        self.send_data(0xf7)		#border floating	

        self.send_command(0X02)  	#power off
        self.ReadBusy() #waiting for the electronic paper IC to release the idle signal
        self.send_command(0X07)  	#deep sleep
        self.send_data(0xA5)
        
        epdconfig.delay_ms(2000)
        epdconfig.module_exit()
### END OF FILE ###

File: Arduino/epd4in2b_V2/epd4in2b_V2.cpp
/**
 *  @filename   :   epd4in2b_V2.cpp
 *  @brief      :   Implements for Dual-color e-paper library
 *  @author     :	Yehui from Waveshare
 *
 *  Copyright (C) Waveshare     Nov 25 2020
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documnetation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to  whom the Software is
 * furished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <stdlib.h>
#include "epd4in2b_V2.h"

Epd::~Epd() {
};

Epd::Epd() {
    flag=0;
    reset_pin = RST_PIN;
    dc_pin = DC_PIN;
    cs_pin = CS_PIN;
    busy_pin = BUSY_PIN;
    width = EPD_WIDTH;
    height = EPD_HEIGHT;
};

int Epd::Init_old(void) {
    Reset();
    SendCommand(0x04);
    ReadBusy();
	
    SendCommand(0x00);
    SendData(0x0F);     // LUT from OTP
	
    // /* EPD hardware init end */
    return 0;
}

int Epd::Init_new(void) {

    Reset();

    ReadBusy();
    SendCommand(0x12);
    ReadBusy();

    SendCommand(0x3C); //BorderWavefrom
    SendData(0x05);	

    SendCommand(0x18); //Read built-in temperature sensor
    SendData(0x80);	

    SendCommand(0x11); //data entry mode       
    SendData(0x03);

    SendCommand(0x44); //set Ram-X address start/end position   
    SendData(0x00);
    SendData(width/8-1);

    SendCommand(0x45); //set Ram-Y address start/end position          
    SendData(0x00);
    SendData(0x00); 
    SendData((height-1)%256);    
    SendData((height-1)/256);

    SendCommand(0x4E);   // set RAM x address count to 0;
    SendData(0x00);
    SendCommand(0x4F);   // set RAM y address count to 0X199;    
    SendData(0x00);    
    SendData(0x00);
    ReadBusy();

    return 0;
}

int Epd::Init(void) {
    unsigned char i=0;

    /* this calls the peripheral hardware interface, see epdif */
    if (IfInit() != 0) {
        return -1;
    }
    /* EPD hardware init start */
    Reset();

    EPD_GPIO_Init();
    digitalWrite(dc_pin, 0);
    EPD_SendData(0x2F);
    DelayMs(50);
    
    digitalWrite(dc_pin, 1);
    i = EPD_ReadData();
    printf("%02x\n",i);
    
    EPD_SPI_Init();
    if(i == 0x01)
    {
        flag = 0;
        Init_new();
    }
    else
    {
        flag = 1;
        Init_old();
    }
	
    // /* EPD hardware init end */
    return 0;
}

/**
 *  @brief: basic function for sending commands
 */
void Epd::SendCommand(unsigned char command) {
    DigitalWrite(dc_pin, LOW);
    SpiTransfer(command);
}

/**
 *  @brief: basic function for sending data
 */
void Epd::SendData(unsigned char data) {
    DigitalWrite(dc_pin, HIGH);
    SpiTransfer(data);
}

/**
 *  @brief: Wait until the busy_pin goes HIGH
 */
void Epd::ReadBusy(void) {
    if(flag == 0)
    {
        while(DigitalRead(busy_pin) == 1) {
            DelayMs(100);
        } 
    }
    else
    {
        while(DigitalRead(busy_pin) == 0) {
            DelayMs(100);
        } 
    }     
}

/**
 *  @brief: module reset. 
 *          often used to awaken the module in deep sleep, 
 *          see Epd::Sleep();
 */
void Epd::Reset(void) {
    DigitalWrite(reset_pin, HIGH);
    DelayMs(200);   
    DigitalWrite(reset_pin, LOW);
    DelayMs(2);
    DigitalWrite(reset_pin, HIGH);
    DelayMs(200);   
}

/**
 *  @brief: transmit partial data to the black part of SRAM
 */
void Epd::Display_Window_Black(const UBYTE *image, UBYTE count) {
    UBYTE k;
    if(count == 0 && flag == 0)
        SendCommand(0x24);
    else if(count == 0)
        SendCommand(0x10);
    
    for (UWORD j = 0; j < 100; j++) {
        for (UWORD i = 0; i < 50; i++) {
            if(i<16)
            {
                SendData(image[i + (j*16)]);
            }
            else
            {
                SendData(0xFF);
            } 
        }
    }
}

/**
 *  @brief: transmit partial data to the red part of SRAM
 */
void Epd::Display_Window_Red_new(const UBYTE *image, UBYTE count) {
    UBYTE k;
    if(count == 0)
        SendCommand(0x26);
    
    for (UWORD j = 0; j < 100; j++) {
        for (UWORD i = 0; i < 50; i++) {
            if(i<16)
            {
                SendData(~image[i + (j*16)]);
            }
            else
            {
                SendData(0x00);
            } 
        }
    }
}
void Epd::Display_Window_Red_old(const UBYTE *image, UBYTE count) {
    UBYTE k;
    if(count == 0)
        SendCommand(0x13);
    
    for (UWORD j = 0; j < 100; j++) {
        for (UWORD i = 0; i < 50; i++) {
            if(i<16)
            {
                SendData(image[i + (j*16)]);
            }
            else
            {
                SendData(0xFF);
            } 
        }
    }
}
void Epd::Display_Window_Red(const UBYTE *image, UBYTE count) {
    if(flag == 0)
        Display_Window_Red_new(image, count);
    else
        Display_Window_Red_old(image, count);
}

void Epd::DisplayFrame(void) {
    if(flag == 0)
    {
        SendCommand(0x22);
        SendData(0xF7);
        SendCommand(0x20);
        ReadBusy();
    }
    else
    {
        SendCommand(0x12);
        DelayMs(100);
        ReadBusy();
    }
}

/**
 * @brief: refresh and displays the frame
 */
void Epd::Display_old(const UBYTE *blackimage, const UBYTE *ryimage) {
    UBYTE k;
    SendCommand(0x10);
    for (UWORD j = 0; j < height; j++) {
        for (UWORD i = 0; i < width/8; i++) {
          SendData(pgm_read_byte(&blackimage[i + (j*width/8)]));
        }
    }
    
    SendCommand(0x13);
    for (UWORD j = 0; j < height; j++) {
        for (UWORD i = 0; i < width/8; i++) {
          SendData(pgm_read_byte(&ryimage[i + (j*width/8)]));
        }
    }
    SendCommand(0x12);
    DelayMs(100);
    ReadBusy();
}

void Epd::Display_new(const UBYTE *blackimage, const UBYTE *ryimage) {
    UBYTE k;
    SendCommand(0x24);
    for (UWORD j = 0; j < height; j++) {
        for (UWORD i = 0; i < width/8; i++) {
          SendData(pgm_read_byte(&blackimage[i + (j*width/8)]));
        }
    }
    
    SendCommand(0x26);
    for (UWORD j = 0; j < height; j++) {
        for (UWORD i = 0; i < width/8; i++) {
          k = pgm_read_byte(&ryimage[i + (j*width/8)]);
          SendData(~k);
        }
    }
    SendCommand(0x22);
    SendData(0xF7);
    SendCommand(0x20);
    ReadBusy();
}

void Epd::Display(const UBYTE *blackimage, const UBYTE *ryimage) {
    if(flag == 0)
        Display_new(blackimage, ryimage);
    else
        Display_old(blackimage, ryimage);
}

/**
 * @brief: clear the frame data from the SRAM, this won't refresh the display
 */
void Epd::Clear_new() {
    //send black data
    SendCommand(0x24);
    for (UWORD j = 0; j < height; j++) {
        for (UWORD i = 0; i < width/8; i++) {
            SendData(0xff);
        }
    }
    //send red data
    SendCommand(0x26);
    for (UWORD j = 0; j < height; j++) {
        for (UWORD i = 0; i < width/8; i++) {
            SendData(0x00);
        }
    }
    SendCommand(0x22);
    SendData(0xF7);
    SendCommand(0x20);
    ReadBusy();
}

void Epd::Clear_old() {
    //send black data
    SendCommand(0x10);
    for (UWORD j = 0; j < height; j++) {
        for (UWORD i = 0; i < width/8; i++) {
            SendData(0xff);
        }
    }
    //send red data
    SendCommand(0x13);
    for (UWORD j = 0; j < height; j++) {
        for (UWORD i = 0; i < width/8; i++) {
            SendData(0xff);
        }
    }
    SendCommand(0x12);
    DelayMs(100);
    ReadBusy();
}

void Epd::Clear() {
    if(flag == 0)
        Clear_new();
    else
        Clear_old();
}

/**
 * @brief: After this command is transmitted, the chip would enter the deep-sleep mode to save power. 
 *         The deep sleep mode would return to standby by hardware reset. The only one parameter is a 
 *         check code, the command would be executed if check code = 0xA5. 
 *         You can use Epd::Reset() to awaken and use Epd::Init() to initialize.
 */
void Epd::Sleep_new(void) {
    SendCommand(0x10);
    SendData(0x01);
}

void Epd::Sleep_old() {
   SendCommand(0X50);  	
   SendData(0xf7);
   SendCommand(0x02); 
   ReadBusy();
   SendCommand(0x07);  
   SendData(0XA5);
}

void Epd::Sleep(void) {
    if(flag == 0)
        Sleep_new();
    else
        Sleep_old();
}


/* END OF FILE */
